libsql-sqlite3/src/vectordiskann.c

```diff
@@ -0,0 +1,345 @@
+    assert( VECTOR_NODE_METADATA_SIZE + pIndex->nNodeVectorSize + iEdge * pIndex->nEdgeVectorSize < offset );
+    vectorInitStatic(
+      pVector,
+      pIndex->nEdgeVectorType,
+      pBlobSpot->pBuffer + VECTOR_NODE_METADATA_SIZE + (iEdge + 1) * pIndex->nNodeVectorSize,
+      pIndex->nEdgeVectorSize
+    );
+  }
+}
+
+int nodeBinEdgeFindIdx(const DiskAnnIndex *pIndex, const BlobSpot *pBlobSpot, u64 nRowid) {
+  int i, nEdges = nodeBinEdges(pIndex, pBlobSpot);
+
+  for(i = 0; i < nEdges; i++){
+    u64 edgeId;
+    nodeBinEdge(pIndex, pBlobSpot, i, &edgeId, NULL);
+    if( edgeId == nRowid ){
+      return i;
+    }
+  }
+  return -1;
+}
+
+void nodeBinPruneEdges(const DiskAnnIndex *pIndex, BlobSpot *pBlobSpot, int nPruned) {
+  assert( 0 <= nPruned && nPruned <= nodeBinEdges(pIndex, pBlobSpot) );
+
+  writeLE16(pBlobSpot->pBuffer + sizeof(u64), nPruned);
+}
+
+void nodeBinInsertEdge(const DiskAnnIndex *pIndex, BlobSpot *pBlobSpot, int iInsert, u64 nRowid, Vector *pVector) {
+  int nMaxEdges = nodeEdgesMaxCount(pIndex);
+  int nEdges = nodeBinEdges(pIndex, pBlobSpot);
+  int edgeVectorOffset, edgeMetaOffset, itemsToMove;
+
+  assert( 0 <= iInsert && iInsert < nMaxEdges );
+  assert( 0 <= iInsert && iInsert <= nEdges );
+
+  if( nEdges < nMaxEdges ){
+    nEdges++;
+  }
+
+  itemsToMove = nEdges - iInsert - 1;
+  edgeVectorOffset = VECTOR_NODE_METADATA_SIZE + pIndex->nNodeVectorSize + iInsert * pIndex->nEdgeVectorSize;
+  edgeMetaOffset = nodeEdgesMetadataOffset(pIndex) + iInsert * VECTOR_EDGE_METADATA_SIZE;
+
+  assert( edgeVectorOffset + pIndex->nEdgeVectorSize * (itemsToMove + 1) <= pBlobSpot->nBufferSize );
+  assert( edgeMetaOffset + VECTOR_EDGE_METADATA_SIZE * (itemsToMove + 1) <= pBlobSpot->nBufferSize );
+
+  memmove(pBlobSpot->pBuffer + edgeVectorOffset + pIndex->nEdgeVectorSize, pBlobSpot->pBuffer + edgeVectorOffset, itemsToMove * pIndex->nEdgeVectorSize);
+  memmove(pBlobSpot->pBuffer + edgeMetaOffset + VECTOR_EDGE_METADATA_SIZE, pBlobSpot->pBuffer + edgeMetaOffset, itemsToMove * VECTOR_EDGE_METADATA_SIZE);
+
+  vectorSerializeToBlob(pVector, pBlobSpot->pBuffer + edgeVectorOffset, pIndex->nEdgeVectorSize);
+  writeLE64(pBlobSpot->pBuffer + edgeMetaOffset + sizeof(u64), nRowid);
+
+  writeLE16(pBlobSpot->pBuffer + sizeof(u64), nEdges);
+}
+
+void nodeBinDeleteEdge(const DiskAnnIndex *pIndex, BlobSpot *pBlobSpot, int iDelete) {
+  int nEdges = nodeBinEdges(pIndex, pBlobSpot);
+  int edgeVectorOffset, edgeMetaOffset, itemsToMove;
+
+  assert( 0 <= iDelete && iDelete < nEdges );
+
+  itemsToMove = nEdges - iDelete - 1;
+  edgeVectorOffset = VECTOR_NODE_METADATA_SIZE + pIndex->nNodeVectorSize + iDelete * pIndex->nEdgeVectorSize;
+  edgeMetaOffset = nodeEdgesMetadataOffset(pIndex) + iDelete * VECTOR_EDGE_METADATA_SIZE;
+
+  assert( edgeVectorOffset + pIndex->nEdgeVectorSize * (itemsToMove + 1) <= pBlobSpot->nBufferSize );
+  assert( edgeMetaOffset + VECTOR_EDGE_METADATA_SIZE * (itemsToMove + 1) <= pBlobSpot->nBufferSize );
```

Comment:
This is probably wrongly copy pasted from nodeBinInsertEdge. We can't break this invariant when we're deleting edge.

Reply:
Hm...but that's what invariants are made for - we should not break them)

Here I wanted to check that we will not access memory outside of the buffer during the shift process (maybe I'm actually better to remove this assertions).

Reply:
Where does the +1 come from when we're deleting? It made sense in the insertion function because we were adding one element but for deletion I can't connect the dots between +1 and the operation. Insertion was shifting left so it was increasing the range by 1 but deletion shifts right so it is only decreasing the range thus it can never break the invariant if it hold true at the start of the function.

Reply:
So the assertion is really checking the invariant before the operation not after but I guess it's fine.